migration_scripts/bulk_load/src/footprint.clj

```diff
@@ -0,0 +1,263 @@
+  [db]
+  (let [partitioned
+        #(fn [[_ part]] [% (edn/read-string part)])
+
+        table->entity
+        (fn [table]
+          (condp re-find table
+            #"^checkpoints" [:checkpoints]
+            #"^display"     [:display]
+            #"^epochs"      [:epochs]
+
+            #"^event_emit_package"         [:event-emit-package]
+            #"^event_emit_module"          [:event-emit-module]
+            #"^event_senders"              [:event-senders]
+            #"^event_struct_package"       [:event-struct-package]
+            #"^event_struct_module"        [:event-struct-module]
+            #"^event_struct_name"          [:event-struct-name]
+            #"^event_struct_instantiation" [:event-struct-instantiation]
+
+            #"^objects(_pkey)?$"           [:objects]
+            #"^objects_snapshot(_pkey)?$"  [:objects-snapshot]
+
+            #"^packages"                   [:packages]
+
+            #"^tx_calls_pkg"               [:tx-calls-pkg]
+            #"^tx_calls_mod"               [:tx-calls-mod]
+            #"^tx_calls_fun"               [:tx-calls-fun]
+            #"^tx_changed_objects"         [:tx-changed-objects]
+            #"^tx_digests"                 [:tx-digests]
+            #"^tx_input_objects"           [:tx-input-objects]
+            #"^tx_kinds"                   [:tx-kinds]
+            #"^tx_recipients"              [:tx-recipients]
+            #"^tx_senders"                 [:tx-senders]
+
+            #"^events_partition_(\d+)" :>>
+            (partitioned :events)
+
+            #"^objects_history_partition_(\d+)" :>>
+            (partitioned :objects-history)
+
+            #"^objects_version_([0-9a-f]{2})" :>>
+            (fn [[_ part]] [:objects-version (Integer/parseInt part 16)])
+
+            #"^transactions_partition_(\d+)" :>>
+            (partitioned :transactions)))]
+    (->> ["SELECT
+               relname,
+               relkind,
+               reltuples,
+               pg_table_size(CAST(relname AS VARCHAR)) tab_size,
+               pg_relation_size(CAST(relname AS VARCHAR)) rel_size
+           FROM
+               pg_class
+           WHERE
+               relkind IN ('i', 'r')
+           AND relname NOT LIKE 'pg_%'
+           AND relname NOT LIKE 'sql_%'
+           AND relname NOT LIKE '__diesel_%'"]
+         (jdbc/execute! db)
+
+         ;; Each row represents a table or an index. Figure out
+         ;; which (potentially partitioned) table it belongs to.
+         (map (fn [{:pg_class/keys [relname relkind reltuples]
+                    :keys [tab_size rel_size]}]
+                [(table->entity relname)
+                 (cond
+                   (= relkind "r")
+                   {:self rel_size
+                    :toast (- tab_size rel_size)
+                    :tuples reltuples}
+                   (s/ends-with? relname "_pkey")
+                   {:pkey rel_size}
+                   :else
+                   {:idx rel_size})]))
```

Comment:
i see, so postgres returns different data for tables and indexes (or rather, index doesn't have a toast and we don't really need the tuple info from index)

Reply:
That's right!